Content.Server/StationEvents/Events/AirlockVirusRule.cs

```diff
@@ -0,0 +1,61 @@
+using Content.Server.StationEvents.Components;
+using Content.Shared.GameTicking.Components;
+using JetBrains.Annotations;
+using Robust.Shared.Player;
+using Robust.Shared.Random;
+using Content.Server.Announcements.Systems;
+using Content.Server.GameTicking;
+using Content.Shared.Emag.Systems;
+using Robust.Shared.Timing;
+using Content.Server.Station.Components;
+using Content.Server.Station.Systems;
+
+namespace Content.Server.StationEvents.Events;
+
+[UsedImplicitly]
+public sealed class AirlockVirusRule : StationEventSystem<AirlockVirusRuleComponent>
+{
+    [Dependency] private readonly AnnouncerSystem _announcer = default!;
+    [Dependency] private readonly GameTicker _gameTicker = default!;
+    [Dependency] private readonly EmagSystem _emag = default!;
+    [Dependency] private readonly IRobustRandom _random = default!;
+    [Dependency] private readonly StationSystem _station = default!;
+
+    protected override void Started(EntityUid uid, AirlockVirusRuleComponent component, GameRuleComponent gameRule, GameRuleStartedEvent args)
+    {
+        base.Started(uid, component, gameRule, args);
+
+        var station = _gameTicker.GetSpawnableStations();
+        if (station is null)
+            return;
+        var stationGrids = new HashSet<EntityUid>();
+        foreach (var stations in station)
+        {
+            if (TryComp<StationDataComponent>(stations, out var data) && _station.GetLargestGrid(data) is { } grid)
+                stationGrids.Add(grid);
+        }
+
+        var query = EntityManager.EntityQueryEnumerator<AirlockVirusTargetComponent>();
+        List<EntityUid> airlocks = new();
+        while (query.MoveNext(out var airlockUid, out var _))
+        {
+            var parent = Transform(airlockUid).GridUid;
+            if (parent is null
+                || !stationGrids.Contains(parent!.Value))
+                continue;
+
+            airlocks.Add(airlockUid);
+        }
+        foreach (var target in airlocks)
+            Timer.Spawn(TimeSpan.FromSeconds(_random.NextDouble(component.MinimumTimeToEmag, component.MaximumTimeToEmag)), () =>
+                _emag.DoEmagEffect(uid, target));
```

Comment:
you can probably remove the whole `airlocks` list and do it inside the `while` block

Reply:
I'm fatally allergic to while loops lmao. I'm horribly foreachpilled.

Reply:
you're iterating over the airlocks twice basically. First time you're gathering them in `while` querry, next time you're applying effects to them. This is simply redundant

Reply:
Ahhh. Actually I see how you're right. gimme a few minutes to drop the PR I'm about to drop and I'll do this right away.

Reply:
Got it! There is now no longer a second loop. If an airlock passes the check, it targets that airlock with the emag timer.